Jenkinsfile

```diff
@@ -0,0 +1,246 @@
+            . .py3venv/bin/activate
+            curl https://bootstrap.pypa.io/get-pip.py | python
+            id
+            python setup.py install
+            pip install coverage mock
+            pip install -I mbed-ls==1.2.14
+            coverage run -m unittest discover -s test -vvv
+            coverage html
+            coverage xml
+            deactivate
+            """
+        } else {
+            // create python2 venv, do installation and run tests
+            sh """
+            virtualenv --python=../usr/bin/python .py2venv --no-site-packages
+            . .py2venv/bin/activate
+            id
+            python setup.py install
+            pip install coverage mock
+            coverage run -m unittest discover -s test -vvv
+            coverage html
+            coverage xml
+            deactivate
+            """
+        }
+        setBuildStatus('SUCCESS', "${buildName}", "test done")
+    } catch (err) {
+        setBuildStatus('FAILURE', "${buildName}", "test failed")
+        currentBuild.result = 'FAILURE'
+    }
+}
+
+
+def linux_pylint_check() {
+    // execute pylint check shell (only for python2, and linux)
+    sh './run_pylint.sh | tee pylint.log'
+
+    // publish warnings
+    step([
+        $class: 'WarningsPublisher',
+        consoleParsers: [
+            [
+                parserName: 'GNU Make + GNU C Compiler (gcc)'
+            ]
+        ],
+        parserConfigurations: [
+            [
+                parserName: 'PyLint',
+                pattern: '**/pylint.log'
+            ]
+        ],
+        unstableTotalAll: '0',
+        failedTotalAll: '0'
+    ])
+}
+
+
+def winTest(String pythonVersion) {
+    echo "set ${pythonVersion} github status"
+    String buildName = "windows ${pythonVersion} test"
+    setBuildStatus('PENDING', "${buildName}", 'test start')
+    try {
+        if (pythonVersion == "py3") {
+            // create python 3 venv
+            echo 'hello windows py3 starts'
+            bat """
+                c:\\Python36\\python.exe -m venv py3venv
+                echo "Activating venv"
+                py3venv\\Scripts\\activate.bat & pip install coverage mock & pip install -I mbed-ls==1.2.14 & python setup.py install & coverage run -m unittest discover -s test -vvv & coverage html & coverage xml & deactivate
+            """
+        } else {
+            // create python2 venv, do installation and run tests
+            echo 'hello windows py2 starts'
+            bat """
+                virtualenv --python=c:\\Python27\\python.exe py2venv --no-site-packages
+                echo "Activating venv"
+                py2venv\\Scripts\\activate.bat & pip install coverage mock & python setup.py install & coverage run -m unittest discover -s test -vvv & coverage html & coverage xml & deactivate
+            """
+        }
+        setBuildStatus('SUCCESS', "${buildName}", "test done")
+    } catch (err) {
+        echo "Caught: ${err}"
```

Comment:
I'd suggest adding a simple `echo "Caught exception: ${err}"` to every other catch block too. It's useful to see exactly where the exception was caught.

Reply:
done